packages/remirror__extension-entity-reference/src/entity-reference-extension.ts

```diff
@@ -0,0 +1,264 @@
+import {
+  ApplySchemaAttributes,
+  command,
+  CommandFunction,
+  CreateExtensionPlugin,
+  EditorState,
+  extension,
+  getTextSelection,
+  Helper,
+  helper,
+  Mark,
+  MarkExtension,
+  MarkExtensionSpec,
+  MarkSpecOverride,
+  PrimitiveSelection,
+  ProsemirrorNode,
+  Transaction,
+  uniqueId,
+  within,
+} from '@remirror/core';
+import { Node } from '@remirror/pm/model';
+import { DecorationSet } from '@remirror/pm/view';
+
+import {
+  ActionType,
+  EntityReferenceMetaData,
+  EntityReferenceOptions,
+  EntityReferencePluginState,
+} from './types';
+import { decorateEntityReferences } from './utils/decorate-entity-references';
+import { getDisjoinedEntityReferencesFromNode } from './utils/disjoined-entity-references';
+import { joinDisjoinedEntityReferences } from './utils/joined-entity-references';
+
+/**
+ *  Required props to create entityReference marks decorations.
+ */
+interface StateProps {
+  extension: EntityReferenceExtension;
+  state: EditorState;
+}
+
+const getEntityReferencesFromPluginState = (props: StateProps) => {
+  const { extension, state } = props;
+  const { entityReferences } = extension.pluginKey.getState(state) as EntityReferencePluginState;
+  return entityReferences;
+};
+
+/**
+ * Creates a decoration set from the passed through state.
+ *
+ * @param props - see [[`DecorationProps`]]
+ */
+const createDecorationSet = (props: StateProps) => {
+  const { extension, state } = props;
+  const entityReferences = getEntityReferencesFromPluginState(props);
+  return DecorationSet.create(state.doc, extension.options.getStyle(entityReferences));
+};
+
+@extension<EntityReferenceOptions>({
+  defaultOptions: {
+    getStyle: decorateEntityReferences,
+    blockSeparator: undefined,
+    createId: uniqueId,
+  },
+})
+export class EntityReferenceExtension extends MarkExtension<EntityReferenceOptions> {
+  get name(): string {
+    return 'entity-reference' as const;
+  }
+
+  createMarkSpec(extra: ApplySchemaAttributes, override: MarkSpecOverride): MarkExtensionSpec {
+    const idAttr = `data-entityReference`;
+
+    return {
+      ...override,
+      excludes: '',
+      attrs: {
+        ...extra.defaults(),
+        id: { default: '' },
+      },
+      // Note that we don't implement `ParseDom`.
+      // The reasoning behind this is to disable pasting (parsing) the entityReference marks in the editor.
```

Comment:
I think we should `parseDOM` by default - for consistency with every other extension, but give the user to opt out.

```ts
const spec = {
  // stuff here
};
if (!this.option.disableParseDOM) { // double negative - maybe confusing?
  spec.parseDOM = [
    {
      tag: `span[${idAttr}]`,
      getAttrs: (node) => {
        if (!isElementDomNode(node)) {
          return false;
        }

        const id = node.getAttribute(idAttr);
        return { ...extra.parse(node), id };
      },
    },
    ...(override.parseDOM ?? []),
  ]
}
return spec
```

Reply:
A better is actually to always have the `parseDom` in the `EntityReferenceExtension`.
For our app, we can just inherit the class, and overwrite `createMarkSpec` accordingly.
This way, the disabling doesn't leak into the Remirror extension itself.